File: recipes/subms-health/java/src/main/java/com/submillisecond/recipes/health/SampleApp.java
package com.submillisecond.recipes.health;

import java.util.concurrent.atomic.AtomicBoolean;

/**
 * Sample app: a tour of {@code subms-health} over a trading gateway's
 * {@code /health}. Its readiness rolls up from a market-data feed, a pre-trade
 * risk check, and its exchange session, plus a non-critical order-book cache and
 * a redacted deploy section. Run:
 * {@code mvn -q compile exec:java -Dexec.mainClass=com.submillisecond.recipes.health.SampleApp}
 *
 * <p>The Java port has no optional-feature classes (the Rust {@code datetime} /
 * {@code otel} Cargo features have no Java equivalent), so this is the base tour:
 * worst-wins rollup, probe-aware HTTP codes, non-critical demotion, and secret
 * redaction.
 */
public final class SampleApp {

    public static void main(String[] args) {
        gatewayReadiness();
    }

    static void gatewayReadiness() {
        System.out.println("== base: trading gateway readiness ==");

        AtomicBoolean sessionUp = new AtomicBoolean(true);
        AtomicBoolean riskTight = new AtomicBoolean(false);

        // Sync config: no background threads, we drive refreshNow() ourselves.
        HealthRegistry reg = new HealthRegistry(HealthConfig.sync());

        reg.registerFn("market-data-feed",
                () -> ComponentHealth.up().withDetail("msgs_per_sec", 184_000L),
                new RefreshPolicy().kinds(ProbeKind.READINESS));

        reg.registerFn("risk-check",
                () -> riskTight.get()
                        ? ComponentHealth.degraded("pre-trade limit 92% utilised")
                        : ComponentHealth.up().withDetail("limit_utilisation", "0.41"),
                new RefreshPolicy().kinds(ProbeKind.READINESS));

        reg.registerFn("exchange-session",
                () -> sessionUp.get()
                        ? ComponentHealth.up().withDetail("venue", "XLON")
                        : ComponentHealth.down("gateway logged out"),
                new RefreshPolicy().kinds(ProbeKind.LIVENESS, ProbeKind.READINESS));

        // Non-critical: a cold order-book cache never fails readiness, it is
        // demoted to WARN so the gateway keeps serving.
        reg.registerFn("orderbook-cache",
                () -> ComponentHealth.down("warm cache miss"),
                new RefreshPolicy().critical(false));

        EnvProvider.MapEnv env = new EnvProvider.MapEnv()
                .with("KICKSTART_ENV", "prod")
                .with("KICKSTART_REGION", "eu-west-1")
                .with("KICKSTART_API_TOKEN", "live-abc123-do-not-log");
        EnvSection deploy = new EnvSection("deploy")
                .prefix("KICKSTART_").stripPrefixInKey(true).lowercaseKeys(true).redactSecrets();
        reg.register(deploy.intoIndicator(env), new RefreshPolicy().critical(false));

        reg.start();

        // Steady state: the non-critical cache is DOWN but demoted to WARN, so the
        // gateway still serves (HTTP 200) even though overall is WARN.
        HealthRegistry.Result r = reg.render();
        System.out.println("  steady:   overall " + reg.status() + " -> HTTP " + r.code());
        if (reg.status() != HealthStatus.WARN) throw new AssertionError("cache demoted, not failed");
        if (r.code() != 200) throw new AssertionError("still serving");
        if (!r.json().contains("\"region\":\"eu-west-1\"")) throw new AssertionError("plain var kept");
        if (!r.json().contains("\"api_token\":\"***\"")) throw new AssertionError("secret masked");
        if (r.json().contains("live-abc123")) throw new AssertionError("raw secret never rendered");

        // Limits tighten: risk-check goes DEGRADED. Readiness 503s (pull from
        // rotation), liveness stays 200 - a degraded gateway is not restarted.
        riskTight.set(true);
        reg.refreshNow();
        int readyCode = reg.renderReadiness().code();
        int liveCode = reg.renderLiveness().code();
        System.out.println("  degraded: overall " + reg.status()
                + " -> ready " + readyCode + ", live " + liveCode);
        if (reg.status() != HealthStatus.DEGRADED) throw new AssertionError("worst-wins over WARN");
        if (readyCode != 503) throw new AssertionError("pulled from rotation");
        if (liveCode != 200) throw new AssertionError("not restarted");

        // The exchange session drops: DOWN wins worst-wins, and now liveness 503s.
        sessionUp.set(false);
        reg.refreshNow();
        liveCode = reg.renderLiveness().code();
        System.out.println("  down:     overall " + reg.status() + " -> live " + liveCode);
        if (reg.status() != HealthStatus.DOWN) throw new AssertionError("DOWN beats DEGRADED");
        if (liveCode != 503) throw new AssertionError("hard down restarts the pod");

        // Recovery: overall settles back to WARN (cache still cold), serving again.
        sessionUp.set(true);
        riskTight.set(false);
        reg.refreshNow();
        r = reg.render();
        System.out.println("  recover:  overall " + reg.status() + " -> HTTP " + r.code());
        if (r.code() != 200) throw new AssertionError("serving again");

        reg.stop();
    }
}

File: recipes/subms-health/rust/examples/sample_app.rs
//! Sample app: a tour of `subms-health`, base API first, then each optional
//! feature. Run the base with `cargo run --example sample_app`; add
//! `--all-features` (or `--features otel` / `--features datetime`) to see the
//! feature sections light up.
//!
//! The scenario is a trading gateway's `/health`. Its readiness rolls up from a
//! market-data feed, a pre-trade risk check, and its exchange session, plus a
//! non-critical order-book cache and a redacted deploy section.
//!
//! * base     - worst-wins rollup, probe-aware codes, non-critical demotion, redaction
//! * datetime - chrono-backed RFC3339 `refreshed_at` on the system clock
//! * otel     - status flips forwarded to OTEL via the subms-otel bridge

use std::sync::Arc;
use std::sync::atomic::{AtomicBool, Ordering};

use subms_health::{
    ComponentHealth, EnvSection, HealthConfig, HealthRegistry, HealthStatus, MapEnv, ProbeKind,
    RefreshPolicy,
};

fn main() {
    base_gateway_readiness();

    #[cfg(feature = "datetime")]
    datetime_timestamps();

    #[cfg(feature = "otel")]
    otel_status_export();
}

/// Base API: a trading gateway serves `/health` off a cached snapshot. Its
/// readiness is the worst of its parts, the HTTP code is probe-aware (a degraded
/// process is pulled from rotation but not restarted), a non-critical failure is
/// demoted so the gateway keeps serving, and the deploy section masks secrets.
fn base_gateway_readiness() {
    println!("== base: trading gateway readiness ==");

    let session_up = Arc::new(AtomicBool::new(true));
    let risk_tight = Arc::new(AtomicBool::new(false));

    // Sync config: no background threads, we drive refresh_now() ourselves - the
    // lowest-latency serve path.
    let mut reg = HealthRegistry::with_config(HealthConfig::sync());

    let ready_critical = RefreshPolicy::new().with_kinds(&[ProbeKind::Readiness]);
    reg.register_fn("market-data-feed", ready_critical.clone(), || {
        ComponentHealth::up().with_detail("msgs_per_sec", 184_000i64)
    });

    {
        let risk_tight = Arc::clone(&risk_tight);
        reg.register_fn("risk-check", ready_critical.clone(), move || {
            if risk_tight.load(Ordering::Relaxed) {
                ComponentHealth::degraded("pre-trade limit 92% utilised")
            } else {
                ComponentHealth::up().with_detail("limit_utilisation", "0.41")
            }
        });
    }

    {
        let session_up = Arc::clone(&session_up);
        reg.register_fn(
            "exchange-session",
            RefreshPolicy::new().with_kinds(&[ProbeKind::Liveness, ProbeKind::Readiness]),
            move || {
                if session_up.load(Ordering::Relaxed) {
                    ComponentHealth::up().with_detail("venue", "XLON")
                } else {
                    ComponentHealth::down("gateway logged out")
                }
            },
        );
    }

    // Non-critical: a cold order-book cache never fails readiness, it is demoted
    // to WARN so the gateway keeps serving.
    reg.register_fn(
        "orderbook-cache",
        RefreshPolicy::new().critical(false),
        || ComponentHealth::down("warm cache miss"),
    );

    let env = MapEnv::new()
        .with("KICKSTART_ENV", "prod")
        .with("KICKSTART_REGION", "eu-west-1")
        .with("KICKSTART_API_TOKEN", "live-abc123-do-not-log");
    let deploy = EnvSection::new("deploy")
        .prefix("KICKSTART_")
        .strip_prefix_in_key(true)
        .lowercase_keys(true)
        .redact_secrets();
    reg.register(
        Arc::new(deploy.into_indicator(Arc::new(env))),
        RefreshPolicy::new().critical(false),
    );

    reg.start(); // sync mode: warms the cache, no thread spawned

    // Steady state: the non-critical cache is DOWN but demoted to WARN, so the
    // gateway is still serving (HTTP 200) even though the overall status is WARN.
    let (code, body) = reg.render();
    println!(
        "  steady:   overall {} -> HTTP {code}",
        reg.status().as_str()
    );
    assert_eq!(
        reg.status(),
        HealthStatus::Warn,
        "cache demoted, not failed"
    );
    assert_eq!(code, 200);
    assert!(body.contains("\"region\":\"eu-west-1\""), "plain var kept");
    assert!(body.contains("\"api_token\":\"***\""), "secret masked");
    assert!(!body.contains("live-abc123"), "raw secret never rendered");

    // Pre-trade limits tighten: risk-check goes DEGRADED. Readiness pulls the pod
    // from rotation (503), but liveness stays 200 - a degraded gateway must not be
    // restarted while it is still draining orders.
    risk_tight.store(true, Ordering::Relaxed);
    reg.refresh_now();
    let (ready_code, _) = reg.render_readiness();
    let (live_code, _) = reg.render_liveness();
    println!(
        "  degraded: overall {} -> ready {ready_code}, live {live_code}",
        reg.status().as_str()
    );
    assert_eq!(reg.status(), HealthStatus::Degraded, "worst-wins over WARN");
    assert_eq!(ready_code, 503, "pulled from rotation");
    assert_eq!(live_code, 200, "not restarted");

    // The exchange session drops entirely: DOWN wins worst-wins over DEGRADED, and
    // now even liveness 503s - restart the pod.
    session_up.store(false, Ordering::Relaxed);
    reg.refresh_now();
    let (live_code, _) = reg.render_liveness();
    println!(
        "  down:     overall {} -> live {live_code}",
        reg.status().as_str()
    );
    assert_eq!(reg.status(), HealthStatus::Down, "DOWN beats DEGRADED");
    assert_eq!(live_code, 503, "hard down restarts the pod");

    // Recovery: session back, risk within limits. Overall settles back to WARN
    // (the cache is still cold) and the gateway serves again.
    session_up.store(true, Ordering::Relaxed);
    risk_tight.store(false, Ordering::Relaxed);
    reg.refresh_now();
    let (code, _) = reg.render();
    println!(
        "  recover:  overall {} -> HTTP {code}",
        reg.status().as_str()
    );
    assert_eq!(code, 200, "serving again");
}

/// `datetime` feature: the system clock stamps `refreshed_at` via chrono's
/// RFC3339 formatter (the zero-dep default hand-rolls the same civil-date shape).
/// Either way the stamp is UTC RFC3339, `YYYY-MM-DDTHH:MM:SSZ`.
#[cfg(feature = "datetime")]
fn datetime_timestamps() {
    println!("\n== datetime: chrono RFC3339 timestamps ==");
    let mut reg = HealthRegistry::with_config(HealthConfig::sync());
    reg.register_fn("clock-probe", RefreshPolicy::new(), ComponentHealth::up);
    reg.start();

    let (_, body) = reg.render();
    let stamp = extract_refreshed_at(&body);
    println!("  refreshed_at = {stamp}");
    assert!(stamp.contains('T') && stamp.ends_with('Z'), "RFC3339 UTC");
    assert_eq!(stamp.len(), 20, "YYYY-MM-DDTHH:MM:SSZ");
}

#[cfg(feature = "datetime")]
fn extract_refreshed_at(json: &str) -> String {
    let key = "\"refreshed_at\":\"";
    let start = json.find(key).expect("refreshed_at present") + key.len();
    let end = start + json[start..].find('"').expect("closing quote");
    json[start..end].to_string()
}

/// `otel` feature: status transitions fan out through the registry's dispatcher.
/// A plain listener can observe them in-process, and the subms-otel
/// `OtelEventBridge` forwards each one to OpenTelemetry as a counter (a no-op
/// until a meter provider is installed, so it is always safe to attach).
#[cfg(feature = "otel")]
fn otel_status_export() {
    use std::sync::atomic::AtomicUsize;
    use subms_health::{OtelEventBridge, on_status_change};

    println!("\n== otel: forward status flips to OpenTelemetry ==");
    let mut reg = HealthRegistry::with_config(HealthConfig::sync());

    let transitions = Arc::new(AtomicUsize::new(0));
    {
        let transitions = Arc::clone(&transitions);
        reg.add_listener(on_status_change(move |_| {
            transitions.fetch_add(1, Ordering::Relaxed);
        }));
    }
    reg.add_bridge(Arc::new(OtelEventBridge::new()));

    let venue_up = Arc::new(AtomicBool::new(true));
    {
        let venue_up = Arc::clone(&venue_up);
        reg.register_fn("exchange-session", RefreshPolicy::new(), move || {
            if venue_up.load(Ordering::Relaxed) {
                ComponentHealth::up()
            } else {
                ComponentHealth::down("venue disconnect")
            }
        });
    }

    reg.start(); // first rebuild sets the baseline - no spurious events
    assert_eq!(transitions.load(Ordering::Relaxed), 0, "baseline only");

    venue_up.store(false, Ordering::Relaxed);
    reg.refresh_now(); // overall + component both flip UP -> DOWN
    let n = transitions.load(Ordering::Relaxed);
    println!("  observed {n} transitions (also forwarded to OTEL)");
    assert!(n >= 1, "a status flip emits a transition event");
}
